File: flatten-2d-vector/grandyang2.java
//grandyang idea2, using just a ref to vec2d and two int variables to iterate, runtime 3ms, faster than 95%
//不直接转换为一维数组，而是维护两个变量x和y，我们将x和y初始化为0，对于hasNext函数，我们检查当前x是否小于总行数，y是否和当前行的列数相同，如果相同，说明要转到下一行，则x自增1，y初始化为0，若此时x还是小于总行数，说明下一个值可以被取出来，那么在next函数就可以直接取出行为x，列为y的数字，并将y自增1
public class Vector2D implements Iterator<Integer> {
    private List<List<Integer>> v = null;
    private int x,y;
    public Vector2D(List<List<Integer>> vec2d) {
        v = vec2d;
        x = y = 0;
    }

    @Override
    public Integer next() {
        return v.get(x).get(y++);
    }

    @Override
    public boolean hasNext() {
        while (x < v.size() && y == v.get(x).size()) {
            x++; 
            y = 0;
        }
        return x < v.size();
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i = new Vector2D(vec2d);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: serialize-and-deserialize-n-ary-tree/grandyang2.cpp
/*
// Definition for a Node.
class Node {
public:
    int val = NULL;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
/*
Runtime: 288 ms, faster than 65.52% of C++ online submissions for Serialize and Deserialize N-ary Tree.
Memory Usage: 47.2 MB, less than 100.00% of C++ online submissions for Serialize and Deserialize N-ary Tree.
grandyang idea2, level by level
*/

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(Node* root) {
        if (!root) return "#";
        string res;
        queue<Node*> q{{root}};
        while (!q.empty()) {
            Node *t = q.front(); q.pop();
            res += to_string(t->val) + " " + to_string(t->children.size()) + " ";
            for (Node *child : t->children) {
                q.push(child);
            }
        }
        return res;
    }

    // Decodes your encoded data to tree.
    Node* deserialize(string data) {
        istringstream iss(data);
        queue<Node*> qNode;
        queue<int> qSize;
        string val = "", size = "";
        iss >> val;
        if (val == "#") return NULL;
        iss >> size;
        Node *res = new Node(stoi(val), {}), *cur = res;
        qNode.push(cur);
        qSize.push(stoi(size));
        while (!qNode.empty()) {
            Node *t = qNode.front(); qNode.pop();
            int len = qSize.front(); qSize.pop();
            for (int i = 0; i < len; ++i) {
                if (!(iss >> val)) break;
                if (!(iss >> size)) break;
                cur = new Node(stoi(val), {});
                qNode.push(cur);
                qSize.push(stoi(size));
                t->children.push_back(cur);
            }
        }
        return res;
    }
};


// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));
